File: devpulse-backend/routers/collections.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from database import (
    create_collection, get_collections_by_user, get_collection_by_id,
    delete_collection, update_collection
)
from auth import get_current_user as get_current_user_auth, User as DBUser

router = APIRouter()


class CollectionCreate(BaseModel):
    name: str
    description: str = ""
    format: str  # "postman" or "openapi"
    data: str  # JSON string of collection data


class CollectionUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


@router.post("/")
def create_new_collection(req: CollectionCreate, current_user: DBUser = Depends(get_current_user_auth)):
    if req.format not in ("postman", "openapi"):
        raise HTTPException(status_code=400, detail="Format must be 'postman' or 'openapi'")
    result = create_collection(current_user.id, req.name, req.format, req.data, req.description)
    return result


@router.get("/")
def list_collections(current_user: DBUser = Depends(get_current_user_auth)):
    return {"collections": get_collections_by_user(current_user.id)}


@router.get("/{collection_id}")
def get_collection(collection_id: str, current_user: DBUser = Depends(get_current_user_auth)):
    col = get_collection_by_id(collection_id)
    if not col or col["user_id"] != current_user.id:
        raise HTTPException(status_code=404, detail="Collection not found or access denied")
    return col


@router.delete("/{collection_id}")
def remove_collection(collection_id: str, current_user: DBUser = Depends(get_current_user_auth)):
    col = get_collection_by_id(collection_id)
    if not col or col["user_id"] != current_user.id:
        raise HTTPException(status_code=404, detail="Collection not found or access denied")
    success = delete_collection(collection_id)
    return {"success": success}


@router.patch("/{collection_id}")
def update_collection_meta(collection_id: str, req: CollectionUpdate, current_user: DBUser = Depends(get_current_user_auth)):
    col = get_collection_by_id(collection_id)
    if not col or col["user_id"] != current_user.id:
        raise HTTPException(status_code=404, detail="Collection not found or access denied")
    result = update_collection(collection_id, name=req.name, description=req.description)
    return result

File: server/api/collections.ts
import { z } from "zod";
import { TRPCError } from "@trpc/server";
import { router, protectedProcedure, editorProcedure } from "../_core/trpc";
import * as db from "../db";
import {
  getOrSetCache,
  CACHE_TTL,
  cacheKeys,
  invalidateUserCache,
} from "../_core/cache";
import { getPlanLimits } from "../payments";

export const collectionsRouter = router({
  create: editorProcedure
    .input(
      z.object({
        name: z.string().min(1).max(255),
        description: z.string().max(1000).optional(),
        format: z.enum(["postman", "openapi"]),
        data: z.any(),
      })
    )
    .mutation(async ({ input, ctx }) => {
      if (!input.data || typeof input.data !== "object") {
        throw new Error("Invalid collection data: must be a JSON object");
      }

      // Plan-limit enforcement: free users are capped at `maxCollections`.
      // Existing collections above the cap are grandfathered — we only
      // block NEW creation. Pro / Enterprise resolve to Infinity.
      const plan = (ctx.user.plan ?? "free") as "free" | "pro" | "enterprise";
      const limits = getPlanLimits(plan);
      if (Number.isFinite(limits.maxCollections)) {
        const existing = await db.getCollectionsByUserId(ctx.user.id);
        if (existing.length >= limits.maxCollections) {
          throw new TRPCError({
            code: "FORBIDDEN",
            message: `Your ${plan} plan allows ${limits.maxCollections} collections. Upgrade at /pricing to add more.`,
          });
        }
      }

      const collection = await db.createCollection(
        ctx.user.id,
        input.name,
        input.format,
        input.data,
        input.description
      );
      await invalidateUserCache(ctx.user.id);
      return collection;
    }),

  list: protectedProcedure
    .input(
      z
        .object({
          page: z.number().int().min(1).default(1),
          pageSize: z.number().int().min(1).max(100).default(20),
        })
        .optional()
    )
    .query(async ({ input, ctx }) => {
      const cacheKey = cacheKeys.userCollections(ctx.user.id);

      const allCollections = await getOrSetCache(
        cacheKey,
        CACHE_TTL.USER_COLLECTIONS,
        () => db.getCollectionsByUserId(ctx.user.id)
      );

      const page = input?.page ?? 1;
      const pageSize = input?.pageSize ?? 20;
      const total = allCollections.length;
      const paginated = allCollections.slice(
        (page - 1) * pageSize,
        page * pageSize
      );
      return {
        collections: paginated.map(c => ({
          id: c.id,
          name: c.name,
          description: c.description,
          format: c.format,
          totalRequests: c.totalRequests,
          createdAt: c.createdAt,
        })),
        total,
        page,
        pageSize,
        totalPages: Math.ceil(total / pageSize),
      };
    }),

  get: protectedProcedure
    .input(z.object({ id: z.string() }))
    .query(async ({ input, ctx }) => {
      const collection = await db.getCollectionById(input.id);
      if (!collection || collection.userId !== ctx.user.id) {
        throw new Error("Collection not found or access denied");
      }
      return collection;
    }),

  delete: editorProcedure
    .input(z.object({ id: z.string() }))
    .mutation(async ({ input, ctx }) => {
      const collection = await db.getCollectionById(input.id);
      if (!collection || collection.userId !== ctx.user.id) {
        throw new Error("Collection not found or access denied");
      }
      await db.deleteCollection(input.id);
      return { success: true };
    }),

  update: editorProcedure
    .input(
      z.object({
        id: z.string(),
        name: z.string().min(1).max(255).optional(),
        description: z.string().max(1000).optional(),
      })
    )
    .mutation(async ({ input, ctx }) => {
      const collection = await db.getCollectionById(input.id);
      if (!collection || collection.userId !== ctx.user.id) {
        throw new Error("Collection not found or access denied");
      }
      await db.updateCollection(input.id, {
        name: input.name,
        description: input.description,
      });
      return { success: true };
    }),

  getWithDetails: protectedProcedure
    .input(z.object({ id: z.string() }))
    .query(async ({ input, ctx }) => {
      const collection = await db.getCollectionById(input.id);
      if (!collection || collection.userId !== ctx.user.id) {
        throw new Error("Collection not found or access denied");
      }

      const [scans, recentFindings, shadowApis, complianceReports] =
        await Promise.all([
          db.getScansByCollectionId(input.id),
          db.getFindingsByScanId(
            (await db.getScansByCollectionId(input.id))[0]?.id || ""
          ),
          db.getShadowAPIsByCollectionId(input.id),
          db.getComplianceReportsByCollectionId(input.id),
        ]);

      const lastScan = scans.length > 0 ? scans[0] : null;
      const totalFindings = scans.reduce(
        (sum, scan) => sum + (scan.totalFindings || 0),
        0
      );

      return {
        id: collection.id,
        name: collection.name,
        description: collection.description,
        format: collection.format,
        totalRequests: collection.totalRequests,
        createdAt: collection.createdAt,
        updatedAt: collection.updatedAt,
        lastScanDate: lastScan?.completedAt || lastScan?.createdAt || null,
        totalScans: scans.length,
        totalFindings,
        recentFindings: recentFindings.slice(0, 10).map(f => ({
          id: f.id,
          title: f.title,
          severity: f.severity,
          status: f.status,
          createdAt: f.createdAt,
        })),
        shadowApis: shadowApis.map(s => ({
          id: s.id,
          endpoint: s.endpoint,
          method: s.method,
          riskLevel: s.riskLevel,
          isDocumented: s.isDocumented,
          createdAt: s.createdAt,
        })),
        complianceReports: complianceReports.map(r => ({
          id: r.id,
          reportType: r.reportType,
          complianceScore: parseFloat(r.complianceScore as any),
          totalRequirements: r.totalRequirements,
          metRequirements: r.metRequirements,
          createdAt: r.createdAt,
        })),
      };
    }),
});
